### PR TITLE
MSC4123: Allow `knock` -> `join` transition

### PR DESCRIPTION
**Pull request by Kladki**
[Rendered](https://github.com/Kladki/matrix-spec-proposals/blob/knock-join-transition/proposals/4123-knock-join-transition.md)

### CODE REVIEW

File: proposals/4123-knock-join-transition.md
Comment (tulir) [inline proposals/4123-knock-join-transition.md:1]: @zecakeh pointed out that these transitions are already allowed according to the [auth rules](https://spec.matrix.org/v1.10/rooms/v11/#authorization-rules)

As long as the user isn't banned, rule 4.3.6 allows joining if `join_rule` is `public` regardless of current membership, and rule 4.3.5.3 does the same for restricted rooms if the authorised via user key is valid.

image

So it seems like [the table](https://spec.matrix.org/v1.10/client-server-api/#mroommember) and [the graph](https://spec.matrix.org/v1.10/client-server-api/#room-membership) in the client-server API are just wrong and can be fixed without an MSC

Comment (tulir) [reply proposals/4123-knock-join-transition.md:1]: And apparently ¼ of the pull request already exists: matrix-org/matrix-spec#1709 (missing parts: knock -> join for public rooms in the graph, and updating the table for both cases)

**Comment by Kladki** [comment]
As per https://github.com/matrix-org/matrix-spec-proposals/pull/4123#discussion_r1554948586

File: proposals/4123-knock-join-transition.md
Comment (FSG-Cat) [reply proposals/4123-knock-join-transition.md:1]: i have to ask if this has been confirmed to be correct behavior so we arent walking straight into another Knock Invite mess. Like im not taking a stance on if this needs a MSC or not just asking that we make sure we arent walking into another mess of that magnitude.

**Comment by Kladki** [comment]
ruma (and hence conduit) already allowed the `knock` -> `join` transition, not sure about synapse and dendrite.

File: proposals/4123-knock-join-transition.md
Comment (Kladki) [reply proposals/4123-knock-join-transition.md:1]: Ruma (and hence Conduit) already allowed the `knock` -> `join` transition, not sure about Synapse and Dendrite.

**Pull request closed without merging**